pkg/target/manager.go

```diff
@@ -0,0 +1,399 @@
+		}
+
+		shoot = matchingShoots[0]
+	} else {
+		return errors.New("neither project nor seed were provided to filter the shoot by")
+	}
+
+	// fetch kubeconfig secret
+	secret := corev1.Secret{}
+	key := types.NamespacedName{
+		Name:      fmt.Sprintf("%s.kubeconfig", shoot.Name),
+		Namespace: shoot.Namespace,
+	}
+
+	if err := gardenClient.Get(ctx, key, &secret); err != nil {
+		return fmt.Errorf("failed to retrieve shoot kubeconfig: %w", err)
+	}
+
+	return m.kubeconfigCache.Write(t, secret.Data["kubeconfig"])
+}
+
+func (m *managerImpl) GardenClient(t Target) (client.Client, error) {
+	t, err := m.getTarget(t)
+	if err != nil {
+		return nil, err
+	}
+
+	if t.GardenName() == "" {
+		return nil, ErrNoGardenTargeted
+	}
+
+	return m.clientForGarden(t.GardenName())
+}
+
+func (m *managerImpl) clientForGarden(name string) (client.Client, error) {
+	for _, g := range m.config.Gardens {
+		if g.Name == name {
+			return m.clientProvider.FromFile(g.Kubeconfig)
+		}
+	}
+
+	return nil, fmt.Errorf("targeted garden cluster %q is not configured", name)
+}
+
+func (m *managerImpl) ProjectClient(t Target) (client.Client, error) {
+	t, err := m.getTarget(t)
+	if err != nil {
+		return nil, err
+	}
+
+	if t.GardenName() == "" {
+		return nil, ErrNoGardenTargeted
+	}
+
+	if t.ProjectName() == "" {
+		return nil, ErrNoProjectTargeted
+	}
+
+	// project clients point to the garden cluster, but should be pinned to
+	// the project namespace; pinning is still TODO
+	return m.GardenClient(t)
+}
+
+func (m *managerImpl) SeedClient(t Target) (client.Client, error) {
+	t, err := m.getTarget(t)
+	if err != nil {
+		return nil, err
+	}
+
+	if t.GardenName() == "" {
+		return nil, ErrNoGardenTargeted
+	}
+
+	if t.SeedName() == "" {
+		return nil, ErrNoSeedTargeted
+	}
+
+	kubeconfig, err := m.kubeconfigCache.Read(t)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.clientProvider.FromBytes(kubeconfig)
+}
+
+func (m *managerImpl) ShootNamespaceClient(t Target) (client.Client, error) {
+	return m.SeedClient(t) // TODO: pre-configure the client to use a certain namespace?
+}
+
+func (m *managerImpl) ShootClusterClient(t Target) (client.Client, error) {
+	t, err := m.getTarget(t)
+	if err != nil {
+		return nil, err
+	}
+
+	if t.GardenName() == "" {
+		return nil, ErrNoGardenTargeted
+	}
+
+	if t.SeedName() == "" && t.ProjectName() == "" {
+		return nil, errors.New("neither project nor seed are targeted")
+	}
+
+	if t.ShootName() == "" {
+		return nil, ErrNoShootTargeted
+	}
+
+	kubeconfig, err := m.kubeconfigCache.Read(t)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.clientProvider.FromBytes(kubeconfig)
+}
+
+func (m *managerImpl) patchTarget(patch func(t *targetImpl) error) error {
+	target, err := m.targetProvider.Read()
+	if err != nil {
+		return err
+	}
+
+	// this is horrible cheating
+	impl, ok := target.(*targetImpl)
+	if !ok {
+		return errors.New("target must be using targetImpl as its underlying type")
+	}
```

Comment:
you need to have setters on the target interface instead, right?

Reply:
I want to prevent users from mucking with the target by accident. They were supposed to get some immutable data structure and were supposed to always refer back to the manager if they want another target. That seems to be a worthwhile design decision to keep the interface and teeth grindingly accept this ...cheat.